[MRG+1] Add a stopping criterion in SGD, based on the score on a validation set

TomDLT opened the pull request:
- Follow up #5036, which implemented a stopping criterion based on the training loss.
- This PR implements a stopping criterion based on the prediction score on a validation set.

In both case, the stopping criterion (and the API) is identical with the one in the MLP classes:
After each epoch, we compute the validation score or the training loss. The optimization stops if there is no improvement twice in a row (i.e. the patience is hard-coded to 2).

To match MLP classes API, I added two parameters:
- `early_stopping` (default False), not well named, which selects if we monitor the validation score (`early_stopping=True`) or the training loss (`early_stopping=False`)
- `validation_fraction` (default 0.1), which selects the split size between training set and validation set.

I also added a new learning rate strategy, `learning_rate='adaptive'`, as found in MLP classes:
The learning rate is kept constant, and is divided by 5 when there is no improvement twice in a row. The optimization stops when the learning rate is too small.

TODO:
- [x] make things work
- [x] add a few tests
- [x] benchmark the GIL access at each epoch
- [x] add a few lines in the doc
- [x] add a whats_new entry
- [x] merge #5036

TomDLT committed:
add early stopping with validation set

TomDLT commented:
Here is a benchmark script to check the effect of accessing the GIL at each epoch.
This GIL access is used to compute the prediction score on the validation set, when `early_stopping=True`.

On my laptop, with `n_jobs=4`:
```
5.32 s ± 61.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
2.23 s ± 45.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
2.35 s ± 50 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

<details>

```python
from IPython import get_ipython

import numpy as np

from sklearn.linear_model import SGDRegressor
from sklearn.datasets import make_regression
from sklearn.externals.joblib import Parallel, delayed

ipython = get_ipython()

X, y = make_regression(n_samples=10000, n_features=500, n_informative=50,
                       n_targets=1, bias=10, noise=3., random_state=42)

validation_fraction = 1. / X.shape[0]


def one_run(early_stopping):
    est = SGDRegressor(validation_fraction=validation_fraction,
                       early_stopping=early_stopping,
                       max_iter=100, tol=-np.inf, shuffle=False,
                       random_state=0)
    est.fit(X, y)


n_runs = 4
n_jobs = 4


def single_thread():
    for _ in range(n_runs):
        one_run(False)


def multi_thread():
    delayed_one_run = delayed(one_run)
    Parallel(n_jobs=n_jobs, backend='threading')(delayed_one_run(False)
                                                 for _ in range(n_runs))


def multi_thread_with_GIL_access_at_each_epoch():
    delayed_one_run = delayed(one_run)
    Parallel(n_jobs=n_jobs, backend='threading')(delayed_one_run(True)
                                                 for _ in range(n_runs))


ipython.magic("timeit single_thread()")
ipython.magic("timeit multi_thread()")
ipython.magic("timeit multi_thread_with_GIL_access_at_each_epoch()")
```

</details>

TomDLT commented:
Now that #5036 is merged, is this planned to be in v0.19?

TomDLT commented:
- [ ] Add `n_iter_no_change` parameter, to match `GradientBoosting` API

amueller commented:
related #9456

amueller commented:
I say :-1: for 0.19

jnothman commented:
I'm operating under the assumption that any new features (except perhaps
deprecation) are 👎 for 0.19. We should really be getting 0.19 wrapped up.

On 28 July 2017 at 02:07, Andreas Mueller <notifications@github.com> wrote:

> I say 👎 for 0.19
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9043#issuecomment-318408817>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xO2aqK_xOJKZbGExdQ7qBUlRmeAks5sSLWlgaJpZM4Ny9wT>
> .
>

amueller commented:
Indeed, though we still haven't released a conda-forge package for the RC. though I guess we can move forward with the release without that.

TomDLT committed:
add n_iter_no_change

TomDLT commented:
### Current estimators with early stopping:
- `GradientBoosting(validation_fraction=0.1, n_iter_no_change=None, tol=1e-4)`
   - `n_iter_no_change=None` disables early stopping.
   - `n_iter_no_change=5` enables early stopping based on **validation score**.

- `MLPClassifier(validation_fraction=0.1, early_stopping=False, tol=1e-4)`
   - `early_stopping=True` makes early stopping based on **validation score**.
   - `early_stopping=False` makes early stopping based on **training loss**.
   - `n_iter_no_change` is not a parameter. The value is hard coded and equal to 2.
   - to disables early stopping and force `max_iter`, you need to use `tol=-inf` or `tol=inf`, depending on the early stopping strategy.

- `SGDClassifier(tol=1e-4)`
   - `tol=None` disables early stopping.
   - `tol!=None` enables early stopping based on **training loss**.
   - `n_iter_no_change` is not a parameter. The equivalent value is hard coded and equal to 1.
___
*In this PR*:
- `SGDClassifier(validation_fraction=0.1, early_stopping=False, n_iter_no_change=2, tol=1e-4)`
   - `tol=None` disables early stopping.
   - `n_iter_no_change` is used for both early stopping strategies.
   - `early_stopping=True` makes early stopping based on **validation score**.
   - `early_stopping=False` makes early stopping based on **training loss**.

TomDLT committed:
FIX doctest

TomDLT committed:
Merge branch 'master' into sgd_validation

Conflicts:
	doc/whats_new/v0.20.rst
	sklearn/linear_model/stochastic_gradient.py

jnothman commented on doc/modules/sgd.rst:
```diff
@@ -232,6 +233,27 @@ non-zero attributes per sample.
+The classes :class:`SGDClassifier` and :class:`SGDRegressor` provide two
+criteria to stop the algorithm when a given level of convergence is reached:
+
+  * With ``early_stopping=True``, the input data is splitted into a training
```
splitted -> split

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -44,6 +44,13 @@ Classifiers and regressors
+  :class:`linear_model.PassiveAggressiveRegressor` and
+  :class:`linear_model.Perceptron` now expose a ``early_stopping`` and
+  ``validation_fraction`` parameters, to stop optimization monitoring the
+  score on a validation set. :issue:`9043`  by `Tom Dupre la Tour`_.
```
Add another entry for adaptive learning rate, or put it here. I'm not sure if each estimator needs to be listed. You can reference the user guide instead...?

jnothman commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -36,6 +36,27 @@ class PassiveAggressiveClassifier(BaseSGDClassifier):
+        validation score is not improving by at least tol for
+        n_iter_no_change consecutive epochs.
+
+        .. versionadded:: 0.20
```
How diligent of you to add this :)

jnothman commented on sklearn/linear_model/sgd_fast.pyx:
```diff
@@ -585,6 +627,8 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
     cdef double max_change = 0.0
     cdef double max_weight = 0.0
 
+    cdef short * validation_set_ptr = <short *> validation_set.data
```
I think you can just as well use a typed memoryview above..?

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -233,6 +244,38 @@ def _allocate_parameter_mem(self, n_classes, n_features, coef_init=None,
+        X_train, X_val, y_train, y_val = tmp[:4]
+        idx_train, idx_val, sample_weight_train, sample_weight_val = tmp[4:8]
+
+        self._X_val = X_val
```
should we delattr these at the end of fitting?

TomDLT replied:
Done

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -282,6 +325,8 @@ def fit_binary(est, i, X, y, alpha, C, learning_rate, max_iter,
     penalty_type = est._get_penalty_type(est.penalty)
     learning_rate_type = est._get_learning_rate_type(learning_rate)
 
+    validation_set = est._train_validation_split(X, y, sample_weight)
```
perhaps `validation_mask`?

jnothman commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -817,6 +835,54 @@ def test_multiple_fit(self):
+        clf1 = self.factory(early_stopping=True, random_state=random_state,
+                            validation_fraction=validation_fraction,
+                            learning_rate='constant', eta0=0.01,
+                            tol=None, max_iter=1000, shuffle=shuffle)
```
I don't think it's clear from the documentation what early_stopping=True should do when tol=None

jnothman commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1068,11 +1134,60 @@ def test_partial_fit_equal_fit_optimal(self):
         clf.set_params(epsilon=0.1)
         assert clf.loss_functions['huber'][1] == 0.1
 
+    def test_early_stopping(self):
```
Should we be using inheritance to share these?

jnothman commented on sklearn/utils/seq_dataset.pxd:
```diff
@@ -7,6 +7,7 @@ cimport numpy as np
 
 cdef class SequentialDataset:
     cdef int current_index
```
There are three things called index here. I hope they are documented somewhere.

Also, I'm not sure if we should consider this public interface that's problematic to change... Can't we use `index_data_ptr[current_index]` directly?

TomDLT replied:
Right, I simplified the call using simply `index_data_ptr[current_index]`.
I also added a bit of documentation there.

jnothman reviewed (commented):
This is nice work.

Is it worth adding or modifying an example to show it in action? I know we have an example for gradient boosting's early stopping.

It might be worth adding these frequent parameters to the Glossary. The myriad definitions and implementations of early stopping may also deserve a separate entry as a term.

jnothman commented:
if I understand correctly, this currently will run an extra epoch relative to the current early_stopping=False behaviour?

TomDLT committed:
address various comments

TomDLT committed:
add early stopping example

TomDLT commented:
> if I understand correctly, this currently will run an extra epoch relative to the current early_stopping=False behaviour?

Yes, the default is now `n_iter_no_change=2`, whereas the previous behavior corresponds to `n_iter_no_change=1`. To avoid breaking user's code, we could have the default to 1 and change it in the future...

TomDLT committed:
Merge branch 'master' into sgd_validation

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-c5c0d4c2745a1a6135de443223057aa4991da7e1)

**new alerts:**

* 1 for Encoding error

---

*Comment posted by [lgtm.com](https://lgtm.com)*

TomDLT committed:
add early stopping and n_iter_no_change in the glossary

jnothman commented:
> Yes, the default is now n_iter_no_change=2, whereas the previous behavior corresponds to n_iter_no_change=1. To avoid breaking user's code, we could have the default to 1 and change it in the future...

Will it change anything but `n_iter_` and runtime?

jnothman commented:
I suppose in the boundary case, it could change the issuance of a ConvergenceWarning...

TomDLT commented:
In theory it could make convergence longer, possibly avoiding an early stop and leading to a poorer solution.
___
Also I realize we don't have pandas in CircleC. My example does not desperately need it, but I wonder if we prefer to avoid pandas in examples or if I should simply add it in CircleCI?

TomDLT commented:
![figure_1](https://user-images.githubusercontent.com/11065596/35801324-3153794c-0a6c-11e8-938f-24d8b2552144.png)

jnothman commented:
I think we should avoid pandas in examples that don't illustrate its
integration with scikit-learn...

jnothman reviewed (commented):
I admit it looks nicer with pandas than it would without.

TomDLT committed:
remove pandas from example

jnothman commented:
Just as long without pandas, but slightly less readable!

jnothman commented on doc/modules/sgd.rst:
```diff
@@ -232,6 +233,27 @@ non-zero attributes per sample.
+    the input data.
+
+In both cases, the criterion is evaluated once by epoch, and the algorithm stops
+when the criterion does not improve ``n_iter_no_times`` times in a row. The
```
* n_iter_no_change

jnothman commented on doc/modules/sgd.rst:
````diff
@@ -373,6 +395,11 @@ user via ``eta0`` and ``power_t``, resp.
 to specify the learning rate.
 
+For an adaptively decreasing learning rate, use ``learning_rate='adaptive'``
+and use ``eta0```to specify the starting learning rate. When the stopping
````
You have an extra backtick instead fo a space

jnothman commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,154 @@
+sample. In particular, it is a very efficient method to fit linear models.
+
+As a stochastic method, the loss function is not necessarily decreasing at each
+iteration, and convergence is only guarnateed in expectation. For this reason,
```
*guaranteed

jnothman commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,154 @@
+then fitted on the training set and the stopping criterion is based on the
+prediction score computed on the validation set. This enables us to find the
+least number of iterations which is sufficient to build a model that
+generalizes well to unseen data.
```
and reduces the chance of over-fitting the training data.

jnothman commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,154 @@
+    'Training loss':
+    linear_model.SGDClassifier(early_stopping=False, n_iter_no_change=5,
+                               tol=0.01, random_state=0,
+                               validation_fraction=0.2),
```
This will be unused, right?

jnothman commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,154 @@
+
+This example illustrates how the early stopping can used in the
+:class:`sklearn.linear_model.SGDClassifier` model to achieve
+almost the same accuracy as compared to a model built without early stopping.
```
It might be worth noting that scores differ between the stopping criteria even from early iterations because some of the training data is held out with the validation stopping criterion.

jnothman commented on sklearn/utils/seq_dataset.pyx:
```diff
@@ -15,12 +15,44 @@ np.import_array()
                    int *nnz, double *y, double *sample_weight) nogil:
         """Get the next example ``x`` from the dataset.
 
+        This method get the next sample looping sequentially over all samples.
```
*gets

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -700,17 +765,42 @@ class SGDClassifier(BaseSGDClassifier):
         - 'optimal': eta = 1.0 / (alpha * (t + t0)) [default]
         - 'invscaling': eta = eta0 / pow(t, power_t)
+        - 'adaptive': eta = eta0, as long as the training keeps decreasing.
+        Each time n_iter_no_change consecutive epochs fail to decrease the
```
I don't think this will format right. You need to indent to be under the bullet point. Another option is to reformat this as a definition list (with term on one line and definition indented on the next; no bullets).

jnothman commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -700,17 +765,42 @@ class SGDClassifier(BaseSGDClassifier):
+    validation_fraction : float, optional, default 0.1
+        The proportion of training data to set aside as validation set for
+        early stopping. Must be between 0 and 1.
+        Only used if early_stopping is True
```
Note that this proportion will be taken out of each batch to `partial_fit` (which is a bit counter-intuitive)

jnothman reviewed (commented):
Please add a test to make sure that `partial_fit` works and that small number of samples to `partial_fit` results in non-empty training and validation sets, or else raises an appropriate error.

Otherwise LGTM

Out of curiosity, what do you think of the early stopping API design altogether? It seems to involve a lot of parameters, but I could imagine more: providing a particular CV splitter to handle sample dependencies, providing a particular scorer, etc. Is there a better way to do this? One that would make it easier to extend the features of all the early stopping estimators? I'm not sure we're going to change it now, but it's worth thinking about.

TomDLT committed:
Improve doc and fix typos

TomDLT commented:
Thanks for the review !

> Out of curiosity, what do you think of the early stopping API design altogether? It seems to involve a lot of parameters, but I could imagine more: providing a particular CV splitter to handle sample dependencies, providing a particular scorer, etc. Is there a better way to do this? One that would make it easier to extend the features of all the early stopping estimators? I'm not sure we're going to change it now, but it's worth thinking about.

Good question!
One way to improve it would be to provide a `callback` parameter, which would contain a function called after each epoch, and a `validation_split` function parameter, called before the fit.
```py
def callback(coef, intercept):
    ...
    return stopping_criterion

def validation_split(X, y):
    ...
    return X_train, y_train, X_val, y_val

sgd = SGDClassifier(max_iter=100, callback=callback,
                    validation_split=validation_split)
sgd.fit(X, y)
```

Or we could go all the way to a meta-estimator:
```py
class EarlyStopping(BaseEstimator, MetaEstimatorMixin):
    def __init__(self, estimator, callback, validation_split):
        ...

    def fit(X, y):
        X_train, y_train, X_val, y_val = self.validation_split(X, y)
        for i in range(estimator.max_iter):
             self.estimator.partial_fit(X_train, y_train)
             stopping_criterion = self.callback(X_val, y_val)
             if stopping_criterion:
                 break
```

I have to think a bit more about it...
___
Still TODO:
- [ ] partial_fit

TomDLT committed:
change default to n_iter_no_change=1 to avoid change

TomDLT commented:
About `partial_fit`, I wonder what behavior we want to have.
Indeed, a `partial_fit` is always one and only one full epoch, so how should a stopping criterion actually stop the outer iteration over the different parts of your big dataset?

Should we disable validation split with `partial_fit`?
Or should we store a attribute `convergence_reached_` when the stopping criterion is met?

TomDLT committed:
add check that both sets are not empty in validation split

jnothman commented:
I think raising an error if someone tries to use early stopping and `partial_fit` is a good solution for now.

`validation_split` could be replaced by just getting the first result from a CV splitter, but then we need a way to pass more arguments than X and y in, ideally. But yes, a callback there would help too. In your early stopping callback interface, what is `stopping_criterion` that the callback returns? How do we ensure that the callback has enough information to make a decision? Is stopping the only thing a callback should be able to control (what about adapting learning rates?). We have callbacks in a couple of other places in the library, and users are coming to expect them due to their popular use in other libraries. Do we think this is something we should be aiming to support?

TomDLT commented:
> Is stopping the only thing a callback should be able to control (what about adapting learning rates?)

Good point

> How do we ensure that the callback has enough information to make a decision? 

I would say a callback mostly needs the current coefficients, but we could also add current learning rate, current number of epochs, current loss, ...

> We have callbacks in a couple of other places in the library, and users are coming to expect them due to their popular use in other libraries. 

It seems the only other callback is in [dict_learning](https://github.com/scikit-learn/scikit-learn/blob/32ac22870d004b52b06d12f18ff3ffbf20862b3b/sklearn/decomposition/dict_learning.py#L553) (and sparse-PCA which is based on it). The solution there is to call `callback(locals())`, which is not very clean and might not work in Cython.
___
> Do we think this is something we should be aiming to support?

I have no strong feeling about supporting callbacks in scikit-learn, I guess it would be a nice feature.
For comparison:
- in [scikit-learn-contrib/lightning](https://github.com/scikit-learn-contrib/lightning/blob/e1cdcaab39f4e05c1379feae612e8a7ade9c741c/lightning/impl/sgd_fast.pyx#L412-L416), all solvers have a callback parameter which only returns a stopping signal, and which takes `self` as only argument.
- in [keras/callbacks.py](https://github.com/keras-team/keras/blob/48e1e8458e772758420efba93508de9c03cdf915/keras/callbacks.py#L145) there is an class Callback API to handle different usecases: stop the solver, log the progress, save the intermediate models, schedule the learning rate, visualize models/progress

TomDLT committed:
Merge branch 'master' into sgd_validation

TomDLT committed:
raise error in partial_fit with early_stopping=True

TomDLT committed:
Merge branch 'master' into sgd_validation

Conflicts:
	sklearn/linear_model/sgd_fast.pyx

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 49b43cb85e8c123b36d48a90d7dadbd0a63f0716 into e8d8b8eeb03cbc08247b433a3f5230d900a5f8e8 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-41ff7208010c9fc762b43e83e16d117b31d1e1a0)

**new alerts:**

* 1 for Result of integer division may be truncated

---

*Comment posted by [LGTM.com](https://lgtm.com)*

glemaitre commented on sklearn/linear_model/sgd_fast.pyx:
```diff
@@ -705,13 +756,31 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
+                with gil:
+                    score = estimator._validation_score(weights, intercept)
+                if tol > -INFINITY and score < previous_score + tol:
+                    no_improvement_count += 1
```
What is exactly happening if I have a sequence of: no improvement, improvement, no improvement. I have the impression that we do not reinitialize the improvement counter in this case, isn't it (I might miss the line thought :))

TomDLT replied:
Good point, I fixed it.

glemaitre commented on sklearn/linear_model/sgd_fast.pyx:
```diff
@@ -547,6 +583,9 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
                double alpha, double C,
                double l1_ratio,
                SequentialDataset dataset,
+               np.ndarray[short, ndim=1, mode='c'] validation_mask,
```
unsigned char should be enough?

glemaitre commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -235,6 +248,61 @@ def _allocate_parameter_mem(self, n_classes, n_features, coef_init=None,
+        self._X_val = X_val
+        self._y_val = y_val
+        self._sample_weight_val = sample_weight_val
+        validation_mask = np.zeros(n_samples, dtype=np.int16)
```
np.uint8 should be enough

glemaitre reviewed (commented):
First review before lunch. Basically only nitpicks.

glemaitre reviewed (changes requested):
Otherwise it looks good

TomDLT committed:
address review comments

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -284,6 +352,8 @@ def fit_binary(est, i, X, y, alpha, C, learning_rate, max_iter,
 
+    validation_mask = est._make_validation_split(X, y, sample_weight)
+
     # XXX should have random_state_!
```
the object should store a `self.random_state_` attribute, instead of starting again from `self.random_state` each time.
I think this is unrelated to this PR.

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 685d7073c958d8ed496d9de60943d0d37adf9c1e into e8d8b8eeb03cbc08247b433a3f5230d900a5f8e8 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-64b522a12bded3787c17fea53a48033c94e80613)

**new alerts:**

* 1 for Result of integer division may be truncated

---

*Comment posted by [LGTM.com](https://lgtm.com)*

ogrisel commented on sklearn/linear_model/sgd_fast.pyx:
```diff
@@ -705,13 +756,35 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
+                    no_improvement_count += 1
+                else:
+                    no_improvement_count = 0
+                previous_score = score
```
We should not use the `previous_score` / `previous_loss` but the `best_score` and `best_loss` observed so far to increment or reset the `no_improvement_count` counter.

ogrisel replied:
The following sequences of validation scores should trigger early stopping with a patience of 5:

0.7, 0.8, 0.9, 0.79, 0.87, 0.81, 0.82

0.9 is the best score and there was no improvement for 5 consecutive iterations thereafter even if there has been many fluctuations.

ogrisel commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,155 @@
+    for this_column in np.unique(results_dict[columns]):
+        mask = results_dict[columns] == this_column
+        ax.plot(results_dict[index][mask], results_dict[values][mask],
+                label=this_column)
```
This code is way too complicated to read (and debug as apparently it fails on recent versions of matplotlib).

I am in favor or loading the raw results into a pandas dataframe and then 

for criteron, goup_df in df.groupby('criterion'):

ogrisel replied:
I know but I think pandas-based plots is much easier to read and maintain, so i am +1 in favor of using pandas.

ogrisel commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -36,6 +36,27 @@ class PassiveAggressiveClassifier(BaseSGDClassifier):
+
+        .. versionadded:: 0.20
+
+    n_iter_no_change : int, default=1
```
I would put the default to 5 (or maybe even 10). Otherwise, the estimator runs the risk to stop way to early.

TomDLT replied:
I agree, but this would break user's code.

ogrisel replied:
Why? If it was hard-coded to 2, setting the default to 1 will also break compat.

ogrisel replied:
Actually, in master the effective patience is 1, my bad.

I think this is a really bad default value. We should at least issue a `FutureWarning` and say that in the next version it will be set to 5 (or 10).

Or because it can be considered a bug to stop way to early at random I would actually be in favor of changing the default right now to 5 (or 10) and mention that as a bugfix in the change log.

ogrisel commented on examples/linear_model/plot_sgd_early_stopping.py:
```diff
@@ -0,0 +1,155 @@
+results = []
+for estimator_name, estimator in estimator_dict.items():
+    print(estimator_name + ': ', end='')
+    for this_max_iter in range(1, max_iter):
```
instead of fitting 50 model for each criterio I would rather do:

```python
for estimator_name, estimator in estimator_dict.items():
    print(estimator_name + ': ', end='') 
    for random_state in range(5):
        for this_max_iter in [1, 5, 10, 15, 20, 30]:
            print('.', end='')
            sys.stdout.flush()

            estimator.set_params(random_state=random_state)
            fit_time, n_iter, train_score, test_score = fit_and_score(
                estimator, this_max_iter, X_train, X_test, y_train, y_test)

            results.append((estimator_name, this_max_iter, random_state,
                            fit_time, n_iter, train_score, test_score))
     print('')

TomDLT replied:
But then we don't see the oscillations of the train and test score.

ogrisel replied:
Alright. That's a good point.

ogrisel reviewed (changes requested):
The problem highlighted in https://github.com/scikit-learn/scikit-learn/pull/9043/files#r198497948 must be fixed: patience is to be computed versus the best-observed train loss / validation score so far.

TomDLT committed:
change previous_* into best_*, and add back pandas in example

ogrisel commented:
The current state of the matplotlib rendering of the example is as follows:

![image](https://user-images.githubusercontent.com/89061/41980767-9583afce-7a27-11e8-9141-557ac3d141f6.png)

https://26334-843222-gh.circle-artifacts.com/0/doc/auto_examples/linear_model/plot_sgd_early_stopping.html#sphx-glr-auto-examples-linear-model-plot-sgd-early-stopping-py

This is why I would like to use simple pandas plots with on groupbys. The current code is too brittle and way too complicated to debug.

TomDLT committed:
add future warning for n_iter_no_change

TomDLT commented:
- I changed `previous_*` into `best_*` as suggested.
- I added a future warning to `n_iter_no_change`.
- I changed the example to use pandas as before.

The example now gives:
![figure_1](https://user-images.githubusercontent.com/11065596/41983400-4609fef6-7a2e-11e8-8581-3b46001f71a5.png)

glemaitre commented:
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}                                                                                                                                                                          Sent from my phone - sorry to be brief and potential misspell.                                                                                                                                                                            From: notifications@github.comSent: 27 June 2018 5:20 pmTo: scikit-learn@noreply.github.comReply to: reply@reply.github.comCc: g.lemaitre58@gmail.com; comment@noreply.github.comSubject: Re: [scikit-learn/scikit-learn] [MRG+1] Add a stopping criterion in SGD, based on the score on a validation set (#9043)  I changed previous_* into best_* as suggested.I added a future warning to n_iter_no_change.I changed the example to use pandas as before.
The example now gives:


—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.

glemaitre commented:
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;} Looks good.                                                                                                                                                                         Sent from my phone - sorry to be brief and potential misspell.

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 5c2f33bf288df7d7e5d07719fad9ddbf93954b9f into 3b5abf76597ce6aff76192869f92647c1b5259e7 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-f2211a0d122730cfa67d80f6879133100f3d3b29)

**new alerts:**

* 1 for Result of integer division may be truncated

---

*Comment posted by [LGTM.com](https://lgtm.com)*

TomDLT committed:
FIX use plot in pandas

ogrisel commented:
The train score curve seems to be inconsistent with `n_iter_no_change=3`: I would have expected the training score criterion to stop much earlier.

ogrisel commented:
Because this PR fixes the bug of previous_loss vs best_loss, I think there is no need to add a futurewarning for `n_iter_no_change`: we are already changing the estimator stopping condition by fixing this bug. We just need to document the bug fix on the stopping criterion in the change log

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 28c46a36b2bad1692e3f338205ff310f945db784 into 3b5abf76597ce6aff76192869f92647c1b5259e7 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-632cfed042b14daa2aed8dbfcdc16342bf2fabea)

**new alerts:**

* 1 for Result of integer division may be truncated

---

*Comment posted by [LGTM.com](https://lgtm.com)*

TomDLT commented:
The bug of `previous_loss` vs `best_loss` is not present in master, since it is equivalent to `n_iter_no_change=1`, so the best loss is also the previous loss.
This was only a mistake in this PR.

glemaitre commented:
@ogrisel any other comment?

TomDLT committed:
FIX lgtm warning

ogrisel commented:
I still have the feeling that the current default behavior (or master and the choice of `n_iter_no_change=1` in this PR) is a bug: it can very often lead to premature stopping, especially on small datasets.

If we consider that this is a bug, we can change the default to `n_iter_no_change=5` instead of issuing a `FutureWarning`.

I would be interested in the opinion of others (maybe @jnothman @amueller ?).

ogrisel reviewed (approved):
Other than the decision on FutureWarning and the default value of the patience parameter, LGTM.

ogrisel commented on doc/glossary.rst:
```diff
@@ -1435,6 +1442,11 @@ functions or non-estimator constructors.
+    ``n_iter_no_change``
+        Number of iterations with no improvement to wait before stopping the
+        iterative procedure. It is typically used with :term:`early stopping` to
+        avoid stopping too early.
```
For googlability, we should mention that this parameter is also named "patience" in other libraries.

ogrisel commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -36,6 +36,27 @@ class PassiveAggressiveClassifier(BaseSGDClassifier):
+        .. versionadded:: 0.20
+
+    n_iter_no_change : int, default=1
+        Number of iterations with no improvement to wait before early stopping.
```
For googlability, we should mention that this parameter is also named "patience" in other libraries.

ogrisel replied:
We should also recommend to set it to a large enough value such as 5 or 10 to avoid premature stopping.

jnothman commented:
Optimisation isn't my expertise, but I can offer a software solution:
change the default as a bug fix, but issue a ChangedBehaviorWarning if the
previous default would have stopped earlier than the new default. Too
complicated?​

jnothman commented:
I suppose what I mean is: warn if there is more than 1 iteration with no
change but fewer than 5.​

ogrisel commented:
I am afraid that would make the code too complicated. I think I would rather stick with the FutureWarning that is easier to understand.

jnothman commented:
Well I'm not entirely against just changing it as a bug, I just don't have
a great idea of how problematic the variation might be...​

ogrisel commented:
Here is a test script on a toy dataset:

```python
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.datasets import load_digits
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split


digits = load_digits()

for n_iter_no_change in [1, 5]:
    for seed in range(5):
        print(f"n_iter_no_change={n_iter_no_change}, seed={seed}")
        model = make_pipeline(
            MinMaxScaler(),
            SGDClassifier(max_iter=1000, tol=1e-3,
                          n_iter_no_change=n_iter_no_change, random_state=seed)
        )
        X_train, X_test, y_train, y_test = train_test_split(
            digits.data, digits.target, test_size=0.2, random_state=seed)
        model.fit(X_train, y_train)
        test_acc = model.score(X_test, y_test)
        print(f"n_iter: {model.steps[-1][1].n_iter_}, test acc: {test_acc:0.3f}")
```
results:
```
n_iter_no_change=1, seed=0
n_iter: 11, test acc: 0.922
n_iter_no_change=1, seed=1
n_iter: 12, test acc: 0.967
n_iter_no_change=1, seed=2
n_iter: 11, test acc: 0.925
n_iter_no_change=1, seed=3
n_iter: 12, test acc: 0.900
n_iter_no_change=1, seed=4
n_iter: 12, test acc: 0.953
n_iter_no_change=5, seed=0
n_iter: 46, test acc: 0.958
n_iter_no_change=5, seed=1
n_iter: 64, test acc: 0.972
n_iter_no_change=5, seed=2
n_iter: 34, test acc: 0.922
n_iter_no_change=5, seed=3
n_iter: 48, test acc: 0.950
n_iter_no_change=5, seed=4
n_iter: 50, test acc: 0.969
```

As you can see `n_iter_no_change=1` result in a detrimental premature stopping most of the time.

ogrisel commented:
The effect is even stronger on a small dataset such as iris:


```
random seed: 0
n_iter_no_change: 1,  n_iter: 3, test acc: 0.600
n_iter_no_change: 5,  n_iter: 19, test acc: 0.767
random seed: 1
n_iter_no_change: 1,  n_iter: 5, test acc: 0.900
n_iter_no_change: 5,  n_iter: 18, test acc: 1.000
random seed: 2
n_iter_no_change: 1,  n_iter: 4, test acc: 0.700
n_iter_no_change: 5,  n_iter: 34, test acc: 0.833
random seed: 3
n_iter_no_change: 1,  n_iter: 5, test acc: 0.700
n_iter_no_change: 5,  n_iter: 22, test acc: 0.933
random seed: 4
n_iter_no_change: 1,  n_iter: 5, test acc: 0.833
n_iter_no_change: 5,  n_iter: 24, test acc: 0.867
```

Arguably, iris is probably too small for serious machine learning, especially with stochastic solvers, but still.

TomDLT committed:
Change default n_iter_no_change to 5, as a bug fix

TomDLT committed:
Merge branch 'master' into sgd_validation

ogrisel commented:
For completeness I have also tried on a larger dataset (covertype), and while it's expected that than a large `n_iter_no_change` is not necessary in that case, it does not seem to hurt test accuracy:

```
random seed: 0
n_iter_no_change: 1,  n_iter: 6, test acc: 0.712
n_iter_no_change: 5,  n_iter: 10, test acc: 0.709
random seed: 1
n_iter_no_change: 1,  n_iter: 6, test acc: 0.708
n_iter_no_change: 5,  n_iter: 10, test acc: 0.712
random seed: 2
n_iter_no_change: 1,  n_iter: 6, test acc: 0.708
n_iter_no_change: 5,  n_iter: 10, test acc: 0.711
random seed: 3
n_iter_no_change: 1,  n_iter: 6, test acc: 0.710
n_iter_no_change: 5,  n_iter: 10, test acc: 0.714
random seed: 4
n_iter_no_change: 1,  n_iter: 6, test acc: 0.707
n_iter_no_change: 5,  n_iter: 10, test acc: 0.709
```

ogrisel commented:
Same run on covertype but using early stopping on a 10% validation split:

```
random seed: 0
n_iter_no_change: 1,  n_iter: 5, test acc: 0.710
n_iter_no_change: 5,  n_iter: 13, test acc: 0.711
random seed: 1
n_iter_no_change: 1,  n_iter: 6, test acc: 0.709
n_iter_no_change: 5,  n_iter: 10, test acc: 0.710
random seed: 2
n_iter_no_change: 1,  n_iter: 5, test acc: 0.710
n_iter_no_change: 5,  n_iter: 19, test acc: 0.711
random seed: 3
n_iter_no_change: 1,  n_iter: 4, test acc: 0.708
n_iter_no_change: 5,  n_iter: 22, test acc: 0.712
random seed: 4
n_iter_no_change: 1,  n_iter: 4, test acc: 0.700
n_iter_no_change: 5,  n_iter: 8, test acc: 0.708
```

In this case, `n_iter_no_change=5` is consistently better than `n_iter_no_change=1` despite the size of the dataset.

ogrisel commented:
@jnothman @TomDLT I will go offline before appveyor has completed. Feel free to merge when green. Based on the runs I made I am confident that n_iter_no_change=5 by default is the good / safe choice.

jnothman commented:
very helpful, Olivier!

jnothman reviewed (commented):
@glemaitre, does this still have your +1?

ogrisel commented:
I reported the appveyor failure here: #11438. I believe it's unrelated to this PR.

TomDLT committed:
reformulate bugfix entry
